### ISSUE TITLE
Solutions are not rendering the assigned version of tests

### ISSUE THREAD

**Issue by guygastineau** [exercism/exercism#4213]
Hi,
  As of now when I `exerercism download -t <trackname> -e <exercise-slug>` a problem I already have I will get the newest version of the test.  However, the test displayed with my solution won't change even after I have submitted with the new test in my exercise directory.

I have noticed as a mentor that when I download a students solution it comes with the newer test as well.

I think it would be really nice if we could actually see the test the student used to pass their most recent solution instead of the test they got when they first downloaded the exercise.

A fix on this would probably get us closer to the feature request from #4171 

Thank you for your time!
Guy Gastineau

**Comment by guygastineau** [comment]
I don't know if handling the git_history differently would help or if the testfile should automatically be uploaded by the cli on submission.

Hopefully this isn't hard to implement.

**Comment by kytrinyx** [comment]
The download command should be getting the pinned version of the test, not the newest one.

**Comment by guygastineau** [comment]
So I should keep this up to tack the bug?

**Comment by iHiD** [comment]
Has anyone confirmed this is actually what happens currently and that there is a bug here?

**Comment by guygastineau** [comment]
I updated the test sweet for clock in the ruby track last night and for grains in the bash track this morning using this step:
`exercism download -t <track> -e <exercise-slug>`

**Comment by guygastineau** [comment]
I just realized that I haven't upgraded the cli on my machine in a few weeks.

I can upgrade and report the behavior when I get home from work.

**Comment by iHiD** [comment]
@guygastineau Thanks :)

@kntsoriano In `app/controllers/api/files_controller.rb`, L40 should be using the git_sha. Can you check everywhere else that `Git::ExercismRepo` is used pls.

**Comment by macta** [comment]
I've hit a similar issue - not downloaded an exercise yet - just viewed the readme  in unmentored mode and it seems pinned to the first version of the file that was served. I have updated go template file and regenerated my exercises and they aren't updating. I suspect some server caching bug.

**Comment by guygastineau** [comment]
@macta 

Check out the two feature request issues referenced above. They essentially will provide a fix for what you are wanting.

**Comment by iHiD** [comment]
For clarity, the bug here is that the exercise **isn't** being locked to a git commit when it should be, not that it's being incorrectly locked.

**Comment by guygastineau** [comment]
@ihiD

Thanks for that clarification.
I have been wondering if I should change the name of this issue to make that clearer.

Do you think I should?

**Comment by iHiD** [comment]
@guygastineau Nice idea. I've just done it. Thanks :)

**Comment by mikewalker** [comment]
Just ran into this issue I think. 

What I did:

Started exercise
Submitted PR https://github.com/exercism/c/pull/329
Download exercise again after merge
Completed exercise using the new tests that where downloaded
Submitted Exercise

Results:

Downloaded tests are the new version
Tests shown with the submitted exercise are the old ones.

UUID for exercise '02fd39e409a446b79116db0d75c26a95'

**Comment by rpottsoh** [comment]
On v1 if you `exercism fetch`'ed an exercise that was already downloaded the CLI would only send you the files that were missing.  If you somehow knew that the test suite was updated you would have to delete your copy of the suite before running `fetch`.  This would get you a copy of the latest test suite.  

What I see on v2 is if I call `download` again (as a student) the CLI sends me the latest version of the test suite, overwriting what was already there.  If I view the test suite on the website it is still the previous (old) version.

I think as a mentor, when I download a student's solution via the CLI, I want to receive the test suite that they last received when they `download`ed via the CLI, not always the test suite they received the first time they downloaded the exercise.

Or stated another way.  The test suite the student has in their folder (provided by the CLI), what is seen on the website, and what is provided to a mentor (via the CLI) should always agree.  Is this the goal?  What is the goal?

**Comment by guygastineau** [comment]
I believe that is the goal.

Furthermore, the feature requests related to this would allow students to update what test is pinned to their solution after they are notified of updates.  They wouldn't be required to update the test though.

**Comment by mikewalker** [comment]
This merge appears to be be working correctly although it has caused an interesting side effect for me. An unavoidable side affect though. :)

The exercise that I submitted using the downloaded tests now fails a couple of tests in the pinned version. This is because before the wrong version of the tests was downloaded. I just downloaded again causing the "pinned" version of the tests to get downloaded. (correct behaviour)

So I think the solution at least for the situation I'm in is to post a new iteration that passes both the pinned, and updated tests. :) I've posted a comment to the mentors, that links to this issue to avoid any confusion. Also to let them know that I am going to post a new iteration that passes the pinned tests, just in case a mentor picks up my solution in the meantime. 

The plus side of all this is the everything should be in sync for future exercises now that pinning is working correctly.
